### ISSUE TITLE
Extracting embeddings before activation

### ISSUE THREAD

**Issue by prratadiya** [KevinMusgrave/pytorch-metric-learning#462]
Hello,
Firstly great job with the library! Appreciate how easy it is to integrate other loss functions in our training pipelines.

I am using CosFaceLoss to train a neural network from which I then extract the embeddings. Everything works well. I was curious to know if it would be possible to extract the embeddings prior to them getting activated in the embedding layer? I am new to PyTorch so not sure if this is independent of the library, but would appreciate any help on these lines!

Network sample:

```
class LMCL_Model(nn.Module):
    def __init__(self, input_dim):
        super(LMCL_Model, self).__init__()
        self.layer1 = nn.Linear(input_dim, 30)
        self.layer2 = nn.Linear(30, 15)
        
    def forward(self, x):
        x = F.relu(self.layer1(x))
        x = F.relu(self.layer2(x))
        return x

model     = LMCL_Model(x_train.shape[1])
optimizer = torch.optim.Adam(model.parameters(), lr=0.001)
loss_fn   = CosFaceLoss(embedding_size=15, num_classes=10, margin=0.25)
loss_optimizer = torch.optim.Adam(loss_fn.parameters(), lr=0.001) 
```

embeddings = model(x) returns me an embedding that is strictly non-negative, which I assume is due to it getting ReLu-activated. I want ReLU to be used in the training phase but not while extracting the embedding during inference. Thank you in advance!

**Comment by KevinMusgrave** [comment]
Maybe this will work?

```python
class LMCL_Model(nn.Module):
    # keep your __init__ unchanged
        
    def forward(self, x):
        x = F.relu(self.layer1(x))
        x = self.layer2(x) # don't do ReLU here
        return x

loss_fn   = CosFaceLoss(embedding_size=15, num_classes=10, margin=0.25)
# stack ReLU here
loss_fn = torch.nn.Sequential(torch.nn.ReLU(), loss_fn)
loss_optimizer = torch.optim.Adam(loss_fn.parameters(), lr=0.001) 
```

**Comment by KevinMusgrave** [comment]
Nevermind that won't work because CosFaceLoss takes arguments that ReLU doesn't. So maybe this instead:

```python
class ReLUBeforeLoss(torch.nn.Module):
    def __init__(self, loss_fn):
        super().__init__()
        self.loss_fn = loss_fn

    def forward(self, embeddings, *args, **kwargs):
        embeddings = F.relu(embeddings)
        return self.loss_fn(embeddings, *args, **kwargs)

loss_fn   = CosFaceLoss(embedding_size=15, num_classes=10, margin=0.25)
loss_fn = ReLUBeforeLoss(loss_fn)
```

**Comment by prratadiya** [comment]
Do you think this would be a right way to do it?:

```
class LMCL_Model(nn.Module):
    def __init__(self, input_dim):
        super(LMCL_Model, self).__init__()
        self.layer1 = nn.Linear(input_dim, 30)
        self.layer2 = nn.Linear(30, 15)
        
    def forward(self, x):
        x = F.relu(self.layer1(x))
        features = self.layer2(x)
        x = F.relu(self.layer2(x))
        return x, features
```
and then in the training block:
`y_pred = model(x_train)[0]`

but for embedding extraction:
`embedding = model(test_sample)[1]`

**Comment by KevinMusgrave** [comment]
Yeah that works too :+1:

**Comment by prratadiya** [comment]
Thanks for the response!

Also on a side note, I felt a slight struggle in getting to use these loss functions for first training a classifier, but then also using it to extract embeddings. I had to navigate deep down through comments on previously closed issues. Do you think the repository can benefit from an example notebook in this case (supervised learning + embedding transfer)? I have been able to come up with one, and would be happy to raise a request if it helps!

**Comment by KevinMusgrave** [comment]
> I have been able to come up with one, and would be happy to raise a request if it helps!

Yes please do! :smile:
